packages/core/core/src/services/entity-validator/validators.ts

```diff
@@ -153,6 +156,168 @@ const addUniqueValidator = <T extends yup.AnySchema>(
+     * and may also be nested within another component.
+     *
+     * We construct a query that takes into account the parent's model UID,
+     * dimensions (such as draft and publish state/locale) and excludes the current
+     * content type entity by its ID if provided.
+     */
+    const {
+      model: parentModel,
+      options: parentOptions,
+      id: excludeId,
+    } = componentContext.parentContent;
+
+    const whereConditions: Record<string, any> = {};
+    const isParentDraft = parentOptions && parentOptions.isDraft;
+
+    whereConditions.publishedAt = isParentDraft ? null : { $notNull: true };
+
+    if (parentOptions?.locale) {
+      whereConditions.locale = parentOptions.locale;
+    }
+
+    if (excludeId && !Number.isNaN(excludeId)) {
+      whereConditions.id = { $ne: excludeId };
+    }
+
+    const queryUid = parentModel.uid;
+    const queryWhere = {
+      ...componentContext.pathToComponent.reduceRight((acc, key) => ({ [key]: acc }), {
+        [updatedAttribute.name]: value,
+      }),
+
+      ...whereConditions,
+    };
+
+    // The validation should pass if there is no other record found from the query
+    return !(await strapi.db.query(queryUid).findOne({ where: queryWhere }));
+  };
+
+  const validateUniqueFieldWithinDynamicZoneComponent = async (
+    startOfPath: string
+  ): Promise<boolean> => {
+    if (!componentContext) {
+      return false;
+    }
+
+    const targetComponentUID = model.uid;
+    // Ensure that the value is unique within the dynamic zone in this entity.
+    const countOfValueInThisEntity = (componentContext?.fullDynamicZoneContent ?? []).reduce(
+      (acc, component) => {
+        if (component.__component !== targetComponentUID) {
+          return acc;
+        }
+
+        const updatedValue = component[updatedAttribute.name];
+        return updatedValue === updatedAttribute.value ? acc + 1 : acc;
+      },
+      0
+    );
+
+    if (countOfValueInThisEntity > 1) {
+      // If the value is repeated in the current entity, the validation fails.
+      return false;
+    }
+
+    // Build a query for the parent content type to find all entities in the
+    // same locale and publication state
+    type QueryType = {
+      select: string[];
+      where: {
+        published_at?: { $eq: null } | { $ne: null };
+        id?: { $ne: number };
+        locale?: string;
+      };
+      populate: {
+        [key: string]: {
+          on: {
+            [key: string]: { where: { [key: string]: string | number | boolean } };
+          };
+        };
+      };
+    };
+
+    // Populate the dynamic zone for any components that share the same value
+    // as the updated attribute.
+    const query: QueryType = {
+      select: ['id'],
+      where: {},
+      populate: {
+        [startOfPath]: {
+          on: {
+            [targetComponentUID]: { where: { [updatedAttribute.name]: updatedAttribute.value } },
```

Comment:
Just a side thought, but we could probably optimize a bit the query by only querying the id column from populated components

Reply:
Thanks! Would this do the trick [`31dcab9` (#20687)](https://github.com/strapi/strapi/pull/20687/commits/31dcab9056435b19a3ac7f4891cc1781d480ed1f)

We need access to the `__component` here 
https://github.com/strapi/strapi/blob/31dcab9056435b19a3ac7f4891cc1781d480ed1f/packages/core/core/src/services/entity-validator/validators.ts#L318-L319

so I thought I would have to do:
`select: ['id', '__component'],`

But it seems to be working with just id